data/operon.1
.\" Man page generated from reStructuredText.
.
.TH OPERON 1 "" "" ""
.SH NAME
operon \- command line music tagger
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.nf
\fBoperon\fP [\-\-version] [\-\-help] [\-v | \-\-verbose] <\fIcommand\fP> [<\fIargument\fP>...]
\fBoperon help\fP <\fIcommand\fP>
.fi
.sp
.SH OPTIONS
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-version
Print the program version
.TP
.B \-v\fP,\fB  \-\-verbose
Verbose mode
.UNINDENT
.SH COMMAND-OVERVIEW
.SS Edit Tags
.nf
\fIadd\fP         Add a tag value
\fIremove\fP      Remove a tag value
\fIset\fP         Set a tag and remove existing values
\fIclear\fP       Remove tags
\fIcopy\fP        Copy tags from one file to another
\fIedit\fP        Edit tags in a text editor
\fIfill\fP        Fill tags based on the file path
.fi
.sp
.SS Show file metadata
.nf
\fIlist\fP        List tags
\fIinfo\fP        List file information
\fIprint\fP       Print tags based on the given pattern
.fi
.sp
.SS Edit Embedded Images
.nf
\fIimage\-extract\fP    Extract embedded images
\fIimage\-set\fP        Set embedded image
\fIimage\-clear\fP      Remove embedded images
.fi
.sp
.SS Miscellaneous
.nf
\fItags\fP        List all common tags
\fIhelp\fP        Display help information
.fi
.sp
.SH EDIT TAGS
.SS add
.sp
Add a new tag \fB<tag>\fP with the value \fB<value>\fP to all files.
.sp
operon add [\-h] <tag> <value> <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon add artist \(aqThe Beatles\(aq song1.ogg song2.ogg
.UNINDENT
.SS remove
.sp
Remove all values from the tag \fB<tag>\fP that match either \fB<value>\fP or
the regular expression \fB<pattern>\fP from all files.
.sp
operon remove [\-h] [\-\-dry\-run] <tag> (\-e <pattern> | <value>) <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-dry\-run
Print the results without changing any files
.TP
.BI \-e\fP,\fB  \-\-regexp \ <regexp>
Remove all tag values that match the given regular expression
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon remove artist \(aqThe Beatles\(aq song.ogg
.UNINDENT
.SS set
.sp
Replace all values of the tag \fB<tag>\fP by \fB<value>\fP in all files.
.sp
operon set [\-h] [\-\-dry\-run] <tag> <value> <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-dry\-run
Print the results without changing any files
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon set artist \(aqThe Beatles\(aq song.ogg
.UNINDENT
.SS clear
.sp
Remove all tags that match \fB<tag>\fP or the regular expression \fB<pattern>\fP
from all files. If \fI\-\-all\fP is specified, all known tags will be removed.
.sp
operon clear [\-h] [\-\-dry\-run] (\-a | \-e <pattern> | <tag>) <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-dry\-run
Print the results without changing any files
.TP
.B \-a\fP,\fB  \-\-all
Remove all tags
.TP
.BI \-e\fP,\fB  \-\-regexp \ <regexp>
Remove all tags that match the given regular expression
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon clear \-a song.ogg
.sp
operon clear \-e \(aqmusicbrainz_.*\(aq song.ogg
.sp
operon clear date song.ogg
.UNINDENT
.SS copy
.sp
Copy all tags from the file \fI<source>\fP to \fI<dest>\fP\&. All tags in \fB<dest>\fP
will be preserved. In case the destination format doesn\(aqt support setting a
tag from source, no tags will be copied. To ignore tags that aren\(aqt
supported by the destination format pass \fI\-\-ignore\-errors\fP\&.
.sp
operon copy [\-h] [\-\-dry\-run] [\-\-ignore\-errors] <source> <dest>
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-dry\-run
Print the results without changing any files
.TP
.B \-\-ignore\-errors
Skip tags which the target file does not support
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon copy song.flac song.ogg
.UNINDENT
.SS edit
.sp
Shows all tags in a text editor and will apply any changes made to the text to
the tags. \fIoperon\fP will use the editor specified in the VISUAL or EDITOR
environment variables and if those are not set fall back to \(aqnano\(aq.
.sp
operon edit [\-h] [\-\-dry\-run] <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-dry\-run
Print the results without changing any files
.UNINDENT
.INDENT 0.0
.TP
.B Example:
VISUAL=vi operon edit song.flac
.UNINDENT
.SS fill
.sp
Fill tags based one file paths and a given pattern.
.sp
operon fill [\-h] [\-\-dry\-run] <pattern> <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.B \-\-dry\-run
show changes, don\(aqt apply them
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon fill \-\-dry\-run "<tracknumber>. <title>" "01. Was Ist Ist.flac"
.UNINDENT
.SH SHOW FILE METADATA
.SS list
.sp
Lists all tags, values and a description of each tag in a table.
.sp
operon list [\-h] [\-a] [\-t] [\-c <c1>,<c2>...] <file>
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-a\fP,\fB  \-\-all
Also list programmatic tags
.TP
.B \-t\fP,\fB  \-\-terse
Output is terse and suitable for script processing
.UNINDENT
.INDENT 0.0
.TP
.B \-c, \-\-columns <name>,...
Defines which columns should be printed and in which order
.TP
.B Example:
operon list \-a song.flac
.sp
operon list \-t \-c tag,value song.ogg
.UNINDENT
.SS info
.sp
Lists non\-tag metadata like length, size and format.
.sp
operon info [\-h] [\-t] [\-c <c1>,<c2>...] <file>
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-t\fP,\fB  \-\-terse
Output is terse and suitable for script processing
.UNINDENT
.INDENT 0.0
.TP
.B \-c, \-\-columns <name>,...
Defines which columns should be printed and in which order
.TP
.B Example:
operon info a.ogg
.UNINDENT
.SS print
.sp
Prints information per file built from tag values. The pattern can be
customized by passing a pattern string (See \fBquodlibet\fP(1) for the
pattern format)
.sp
operon print [\-h] [\-p <pattern>] <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.BI \-p\fP,\fB  \-\-pattern \ <pattern>
Use a custom pattern
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon print \-p "<album> \- <artist>" a.ogg
.UNINDENT
.SH EDIT EMBEDDED IMAGES
.SS image\-extract
.sp
Extract all embedded images to the current working directory or the specified
destination directory.
.sp
operon image\-extract [\-h] [\-\-dry\-run] [\-\-primary] [\-d <destination>] <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-\-dry\-run
Print the found images and resulting file paths but don\(aqt save them
.TP
.B \-\-primary
Only extract the primary images for each file
.TP
.BI \-d\fP,\fB  \-\-destination \ <destination>
Save all images to the specified destination
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon image\-extract asong.mp3 anotherone.ogg
.UNINDENT
.SS image\-set
.sp
Set the provided image as primary embedded image and remove all other
embedded images.
.sp
operon image\-set <image\-file> <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon image\-set cover.jpg song.mp3
.UNINDENT
.SS image\-clear
.sp
Remove all embedded images from all specified files.
.sp
operon image\-clear <file>...
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.UNINDENT
.INDENT 0.0
.TP
.B Example:
operon image\-clear song.mp3
.UNINDENT
.SH MISCELLANEOUS
.SS tags
.sp
List all common tags
.sp
operon tags [\-h] [\-a] [\-t] [\-c <c1>,<c2>...]
.INDENT 0.0
.TP
.B \-h\fP,\fB  \-\-help
Display help and exit
.TP
.B \-a\fP,\fB  \-\-all
Also list programmatic tags
.TP
.B \-t\fP,\fB  \-\-terse
Output is terse and suitable for script processing
.UNINDENT
.INDENT 0.0
.TP
.B \-c, \-\-columns <name>,...
Defines which columns should be printed and in which order
.TP
.B Example:
operon tags \-a
.sp
operon tags \-t \-c tag
.UNINDENT
.SS help
.sp
operon help [<command>]
.INDENT 0.0
.TP
.B Example:
operon help list
.UNINDENT
.SH SEE ALSO
.nf
\fBregex\fP(7)
\fBexfalso\fP(1)
\fBquodlibet\fP(1)
.fi
.sp
.\" Generated by docutils manpage writer.
.


NEWS.rst
.. _News:


4.7.0 - Unreleased
------------------

Packaging Changes
 * Quodlibet now uses ``gtksourceview4`` optionally instead of ``gtksourceview3`` :pr:`4377`
 * Appdata is now installed into ``/usr/share/metainfo`` instead of ``/usr/share/appdata`` :pr:`4375`
 * The tests now use use ``appstream`` optionally instead of ``appstream-util`` :pr:`4376`


.. _release-4.6.0:

4.6.0 (2023-08-22) - Ah, perfect, the banner that satisfies everyone
--------------------------------------------------------------------

Packaging Changes
 * Quodlibet now requires Python 3.8+
 * It also now supports Python 3.11
 * libsoup3 is now required (up from version 2)

ℹ️ A highlight of the changes in 4.6.0 is below.
For full details,
see `the 4.6 commit log <https://github.com/quodlibet/quodlibet/compare/release-4.5.0...release-4.6.0>`_.

General
 * Audiofeeds: enable clipboard when adding feed :pr:`4128` (:user:`Nick Boultbee <declension>`)
 * Fix aggregate units on numeric comparisons :pr:`4089` (:user:`Nick Boultbee <declension>`)
 * Fix playlist renaming bug :pr:`4079` (:user:`Nick Boultbee <declension>`)
 * Remove some senf helpers that are more or less the same in the Python 3 stdlib :pr:`4041` (:user:`lazka`)
 * Require Python 3.8 :pr:`3956` (:user:`lazka`)
 * Move to libsoup3 :pr:`4294` (:user:`Nick Boultbee <declension>`)
 * Add Stop command to Control menu :pr:`4213` (:user:`Markus Koller`)
 * Allow disabling of songlist sorting :pr:`4273` (:user:`Nick Boultbee <declension>`)
 * Add human time tags :pr:`4275` (:user:`Nick Boultbee <declension>`)
 * Nicer prefs switches :pr:`4274` (:user:`Nick Boultbee <declension>`)
 * Fix shuffle history :pr:`4255` (:user:`Thomas Leberbauer`)
 * Fix pop-up menu for Podcasts :pr:`4193` (:user:`Nick Boultbee <declension>`)
 * Add CLI arg: --with-pattern :pr:`4166` (:user:`Felicián Németh`)
 * Modernise entire preferences UI 🎨 :pr:`4175` (:user:`Nick Boultbee <declension>`)
 * Allow songlist editing of simple text tags in-place 🆙 :pr:`4173` (:user:`Nick Boultbee <declension>`)
 * Make --rating-up work when song has no previous rating :pr:`4172` (:user:`Felicián Németh`)
 * Don't make the WaitLoadWindow modal :pr:`4171` (:user:`David Greaves`)
 * Prompt on attempting to extend a playlist with DnD'ed tracks :pr:`2746` (:user:`Meriipu`)
 * Stop query date regex from matching > 4 digit year :pr:`4162` (:user:`afontenot`)
 * Add "edit" button to column tag editor window :pr:`4148` (:user:`afontenot`)
 * Track List: Add shortcut Ctrl-Return to the searchbar :pr:`4141` (:user:`Felicián Németh`)
 * Fix crash when using covergrid + default art size :pr:`4151` (:user:`afontenot`)
 * Disable ordering when active order disappears :pr:`4156` (:user:`afontenot`)
 * Improve notif progress bar :pr:`4147` (:user:`Nick Boultbee <declension>`)
 * Playlists: log about potential import failures :pr:`4146` (:user:`Nick Boultbee <declension>`)
 * Fix regression of --print-query :pr:`4142` (:user:`Felicián Németh`)
 * Rename feeds -> podcasts :pr:`3513` :pr:`4139` (:user:`Nick Boultbee <declension>`)
 * Audiofeeds: handle non-feed URLs :pr:`4126` (:user:`Nick Boultbee <declension>`)
 * Fix image size scaling when cover is reloaded :pr:`4115` (:user:`Thomas Leberbauer`)
 * Covergrid using FlowBox :pr:`4101` (:user:`Thomas Leberbauer`)
 * Softer cover art outline :pr:`4096` (:user:`Nick Boultbee <declension>`)
 * Disable editing for Soundcloud tracks :pr:`4097` (:user:`Nick Boultbee <declension>`)
 * Fix lots of browser padding issues :pr:`4066` (:user:`Nick Boultbee <declension>`)
 * Make songs heading entries editable  :pr:`4047` (:user:`Nick Boultbee <declension>`)
 * XSPF formatting :pr:`4028` (:user:`Nick Boultbee <declension>`)
 * Support library watching in the core :pr:`3654` (:user:`Nick Boultbee <declension>`)
 * Write XSPFs with proper namespace (internal v2) :pr:`3984` (:user:`Nick Boultbee <declension>`)
 * Move library: don't force source dir to exist :pr:`3980` (:user:`Nick Boultbee <declension>`)
 * Fix add playlists errors (:bug:`3973`) :pr:`3974` (:user:`TheMelmacian`)
 * Fix display pattern not updating in browsers :pr:`3975` (:user:`Thomas Leberbauer`)


Developer
 * osx: fix installation with setuptools (:user:`Christoph Reiter <lazka>`)
 * osx: update jhbuild (:user:`Christoph Reiter <lazka>`)
 * osx: fix cargo/rustup usage in jhbuild (:user:`Christoph Reiter <lazka>`)
 * osx_bundle: move to 10.13+ (:user:`Christoph Reiter <lazka>`)


Plugins
 * Improve and fix Sonos plugin :pr:`4290` (:user:`Nick Boultbee <declension>`)
 * Tidy export to folder plugin (:user:`Nick Boultbee <declension>`)
 * Unique filters plugin :pr:`4264` (:user:`LoveIsGrief`)
 * playlist export: prefix relative paths with '.' if the path starts with '#' :pr:`4236` (:user:`Felix Krull`)
 * Add a new internal tag: ~json :pr:`4195` (:user:`Felicián Németh`)
 * 🆕 Bookmarks notifications plugin :pr:`4174` (:user:`Nick Boultbee <declension>`)
 * Move listenbrainz plugin to the right location :pr:`4192` (:user:`Felicián Németh`)
 * the sync_to_device plugin shouldn't delete files named 'cover.jpg' :pr:`4083` (:user:`Jan`)
 * Album art downloader: improve GUI :pr:`4106` (:user:`Nick Boultbee <declension>`)
 * Album art downloader: remove Amazon :pr:`4105` (:user:`Nick Boultbee <declension>`)
 * Embed Images plugin: allow choosing a file from disk :pr:`4071` (:user:`Nick Boultbee <declension>`)
 * Add plugin for rich presence in Discord. :pr:`3998` (:user:`Aditi K`)


Translations
 * Thanks to all our translators old and new.
 * These are now listed on the `Weblate QL project <https://hosted.weblate.org/projects/quodlibet/quodlibet/>`_.



.. _release-4.5.0:

4.5.0 (2022-03-29) - What sort of things would you like to know?
----------------------------------------------------------------

Packaging Changes:
 * Quodlibet should now support Python 3.10
 * This will be the last release supporting Windows 7


General:
 * Allow auto-playing added songs in Queue Only mode :pr:`3947` (:user:`afontenot`)

quodlibet/data/operon.bash
# operon(1) completion                             -*- shell-script -*-
# vim: sts=4 sw=4 et

__tags() {
    operon tags -a -t -c tag | tr '\n' ' '
}

_in_array() {
    local i
    for i in "${@:2}"; do
        [[ $1 = "$i" ]] && return
    done
}

_operon() {
    # Assigned variable by _init_completion:
    #   cur    Current argument.
    #   prev   Previous argument.
    #   words  Argument array.
    #   cword  Argument array size.
    local cur prev words cword
    _init_completion -n = || return

    # Operon commands and options
    local opts=(--help --verbose --version)
    local cmds=(add clear copy edit fill help \
                image-clear image-extract image-set \
                info list print remove set tags)

    # Check if a command was entered already
    local command i
    for (( i=0; i < ${#words[@]}-1; i++ )); do
        if _in_array "${words[i]}" "${cmds[@]}"; then
            command=${words[i]}
            break
        fi
    done

    # Completion per command
    if [[ -n $command ]]; then
        case $command in
            add) # <tag> <value> <file> [<files>]
                nargs=0
                for (( j=i+1; j < ${#words[@]}-1; j++ )); do
                    case "${words[j]}" in
                        -*) continue ;;
                        *)  nargs=$((nargs+1)) ;;
                    esac
                done
                if [[ $nargs -eq 0 ]]; then
                    comps=$(__tags)
                    COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                elif [[ $nargs -ge 2 ]]; then
                    compopt -o default
                    COMPREPLY=()
                fi
                return 0
                ;;

            clear) # [--dry-run] [-a | -e <pattern> | <tag>] <file> [<files>]
                case $cur in
                    -*)
                        comps="--all --dry-run --regexp"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        case $prev in
                            -a|--all)
                                compopt -o default
                                COMPREPLY=()
                                ;;
                            -e|--regex)
                                ;;
                            *)
                                nargs=0
                                for (( j=i+1; j < ${#words[@]}-1; j++ )); do
                                    case "${words[j]}" in
                                        -d|--dry-run) continue ;;
                                        -e|--regexp)  continue ;;
                                        -a|--all)     nargs=$((nargs+1)) ;;
                                        *)            nargs=$((nargs+1)) ;;
                                    esac
                                done
                                if [[ $nargs -eq 0 ]]; then
                                    comps=$(__tags)
                                    COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                                elif [[ $nargs -ge 1 ]]; then
                                    compopt -o default
                                    COMPREPLY=()
                                fi
                                ;;
                        esac
                esac
                return 0
                ;;

            copy) # [--dry-run] [--ignore-errors] <source> <dest>
                case $cur in
                    -*)
                        comps="--dry-run --ignore-errors"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        compopt -o default
                        COMPREPLY=()
                        ;;
                esac
                return 0
                ;;

            edit) # [--dry-run] <file>
                case $cur in
                    -*)
                        comps="--dry-run"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        compopt -o default
                        COMPREPLY=()
                        ;;
                esac
                return 0
                ;;

            fill) # [--dry-run] <pattern> <file> [<files>]
                case $cur in
                    -*)
                        comps="--dry-run"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        # no completion for <pattern>
                        nargs=0
                        for (( j=i+1; j < ${#words[@]}-1; j++ )); do
                            case "${words[j]}" in
                                -*) continue ;;
                                *)  nargs=$((nargs+1)) ;;
                            esac
                        done
                        if [[ $nargs -ge 1 ]]; then
                            compopt -o default
                            COMPREPLY=()
                        fi
                        ;;
                esac
                return 0
                ;;

            help) # [<command>]
                COMPREPLY=($(compgen -W "${cmds[*]}" -- "$cur"))
                return 0
                ;;

            image-clear) # <file> [<files>]
                compopt -o default
                COMPREPLY=()
                return 0
                ;;

            image-extract) # [--dry-run] [--primary] [-d <destination>] <file> [<files>]
                case $cur in
                    -*)
                        comps="--destination --dry-run --primary"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        compopt -o default
                        COMPREPLY=()
                        ;;
                esac
                return 0
                ;;

            image-set) # <image-file> <file> [<files>]
                compopt -o default
                COMPREPLY=()
                return 0
                ;;

            info) # [-t] [-c <c1>,<c2>...] <file>
                case $cur in
                    -*)
                        comps="--columns --terse"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        case $prev in
                            -c|--columns)
                                ;;
                            *)
                                compopt -o default
                                COMPREPLY=()
                                ;;
                        esac
                        ;;
                esac
                return 0
                ;;

            list) # [-a] [-t] [-c <c1>,<c2>...] <file>
                case $cur in
                    -*)
                        comps="--all --columns --terse"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        case $prev in
                            -c|--columns)
                                ;;
                            *)
                                compopt -o default
                                COMPREPLY=()
                                ;;
                        esac
                        ;;
                esac
                return 0
                ;;

            print) # [-p <pattern>] <file> [<files>]
                case $cur in
                    -*)
                        comps="--pattern"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        case $prev in
                            -c|--pattern)
                                ;;
                            *)
                                compopt -o default
                                COMPREPLY=()
                                ;;
                        esac
                        ;;
                esac
                return 0
                ;;

            remove) # [--dry-run] <tag> [-e <pattern> | <value>] <file> [<files>]
                case $cur in
                    -*)
                        comps="--dry-run --regexp"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        nargs=0
                        for (( j=i+1; j < ${#words[@]}-1; j++ )); do
                            case "${words[j]}" in
                                -*) continue ;;
                                *)  nargs=$((nargs+1)) ;;
                            esac
                        done
                        if [[ $nargs -eq 0 ]]; then
                            comps=$(__tags)
                            COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        elif [[ $nargs -ge 2 ]]; then
                            compopt -o default
                            COMPREPLY=()
                        fi
                        ;;
                esac
                return 0
                ;;

            set) # [--dry-run] <tag> <value> <file> [<files>]
                case $cur in
                    -*)
                        comps="--dry-run"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                    *)
                        nargs=0
                        for (( j=i+1; j < ${#words[@]}-1; j++ )); do
                            case "${words[j]}" in
                                -*) continue ;;
                                *)  nargs=$((nargs+1)) ;;
                            esac
                        done
                        if [[ $nargs -eq 0 ]]; then
                            comps=$(__tags)
                            COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        elif [[ $nargs -ge 2 ]]; then
                            compopt -o default
                            COMPREPLY=()
                        fi
                        ;;
                esac
                return 0
                ;;

            tags) # [-t] [-c <c1>,<c2>...]
                case $cur in
                    -*)
                        comps="--all --columns --terse"
                        COMPREPLY=($(compgen -W "$comps" -- "$cur"))
                        ;;
                esac
                return 0
                ;;
        esac
    fi

    # Initial completion
    case "$cur" in
        -*)
            COMPREPLY=($(compgen -W "${opts[*]}" -- "$cur"))
            return 0
            ;;
        *)
            COMPREPLY=($(compgen -W "${cmds[*]}" -- "$cur"))
            return 0
            ;;
    esac

} && \
complete -F _operon operon

# ex: filetype=sh
